glommio/src/error.rs

```diff
@@ -3,9 +3,60 @@
 //
 // This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2020 Datadog, Inc.
 //
-use std::fmt;
+use crate::{channels, executor, sync};
+use std::fmt::{self, Debug};
 use std::os::unix::io::RawFd;
 use std::path::PathBuf;
+use thiserror::Error;
+
+pub type Result<T, V> = std::result::Result<T, GlommioError<V>>;
+
+#[derive(Debug)]
+pub enum ResourceType {
+    Semaphore { requested: u64, available: u64 },
+}
+
+impl fmt::Display for ResourceType {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            ResourceType::Semaphore {
+                requested,
+                available,
+            } => f.write_str(&format!(
+                "Semaphore requested {} but only {} available",
+                requested, available
+            )),
+        }
+    }
+}
+
+#[derive(Error, Debug)]
+/// Composite error type to encompass all error types glommio produces.
+pub enum GlommioError<T: Debug + 'static> {
+    #[error("IO error occurred: {0}")]
+    IoError(#[from] std::io::Error),
+
+    #[error("Lock closed error: {0}")]
+    LockClosedError(#[from] sync::LockClosedError),
+
+    #[error("Channel error: {0}")]
+    ChannelError(#[from] channels::ChannelError<T>),
```

Comment:
I would prefer no to have a ChannelError.

What is special about a channel? The errors it can return are WouldBlock / BrokenPipe. That's very similar between a channel and a semaphore.

It seems to me that is is better to have just WouldBlock / BrokenPipe kinds of errors, and specify the source in the error parameters so the user knows where it comes from

Reply:
I think one reason for ChannelError is returning the actual value in the case of an error so the user can do something with it after.. All that said, it would be nice if we didn't have to deal with it, tbh. In fact it's the reason we have an annoying type parameter is because of the channel error.

Reply:
You are right, I overlooked that.

I will think about it a little bit more. I guess ChannelError is still generic enough because it serves all types of channel.

Another option is maybe to have those errors always returning a type, and `Semaphore` can use them but return `()`. Depends on how they would look like.